### PR TITLE
PlonK proof verifier precompile. for BW6-761

### PR DESCRIPTION
**Pull request by wzmuda**
The precompile is going to be called from a solidity contract whose
sole role will be to pass the buffer with serialized proof, vk and
witness. All the verification is done in the precompile using the ususal
Gnark API.

The precompile lives at the address of 0x85, which is the next available
address in our collection of precompiles. The gas cost was estimated at
21000, although it is not backed by any serious calculations.

Also bump Gnark version to v0.9.2.

### CODE REVIEW

File: core/vm/contracts_ec.go [hunk 1]
```diff
@@ -73,6 +83,61 @@ func RunECPrecompiledContract(evm *EVM, precompile ECPrecompiledContract, input
 	return output, suppliedGas, err
 }
 
+// bw6761PlonkProofVerifyPrecompile implements the verification of a given PlonK proof. It is assumed that the proof
+// was generated using BW6-761 curve.
```
Comment (iamrecursion) [inline core/vm/contracts_ec.go]:
```suggestion
// was generated using the BW6-761 curve.
```

File: core/vm/contracts_ec.go [hunk 2]
```diff
@@ -73,6 +83,61 @@ func RunECPrecompiledContract(evm *EVM, precompile ECPrecompiledContract, input
...
+//
+// The input is assumed to encode, in the following order:
+// - a PlonK proof of a circuit,
+// - a verifying key for the proof,
+// - a public witness suitable for verifying the proof.
```
Comment (iamrecursion) [inline core/vm/contracts_ec.go:92]: It is probably worth specifying a bit more about the encoding details. It's all well and good to have example code that generates it, but I would hope to see a byte-level description here.

File: core/vm/contracts_ec.go [hunk 3]
```diff
@@ -73,6 +83,61 @@ func RunECPrecompiledContract(evm *EVM, precompile ECPrecompiledContract, input
...
+// All the above components must be generated for the same circuit definition, over the BW6-761 finite field.
+//
+// Example of serializing the objects into a single input buffer:
+//
+//	    // Assuming proof, verifyingKey and publicWitness were generated properly
+//		var buf bytes.Buffer
+//		proof.WriteTo(&buf)
+//		verifyingKey.WriteTo(&buf)
+//		publicWitness.WriteTo(&buf)
+//	    // at this point buf contains all the objects and can be turned into []byte with buf.Bytes()
```
Comment (iamrecursion) [inline core/vm/contracts_ec.go]:
```suggestion
//	  // Assuming proof, verifyingKey and publicWitness were generated properly
//		var buf bytes.Buffer
//		proof.WriteTo(&buf)
//		verifyingKey.WriteTo(&buf)
//		publicWitness.WriteTo(&buf)
//	  // at this point buf contains all the objects and can be turned into []byte with buf.Bytes()
```

File: params/protocol_params.go [hunk 4]
```diff
@@ -143,11 +143,12 @@ const (
...
-	Bw6761G1AddGas   uint64 = 600   // Price for BW6-761 elliptic curve G1 point addition
-	Bw6761G1MulGas   uint64 = 12000 // Price for BW6-761 elliptic curve G1 point scalar multiplication
-	Bw6761G2AddGas   uint64 = 4500  // Price for BW6-761 elliptic curve G2 point addition
-	Bw6761G2MulGas   uint64 = 55000 // Price for BW6-761 elliptic curve G2 point scalar multiplication
-	Bw6761PairingGas uint64 = 23000 // Price for BW6-761 elliptic curve pairing check
+	Bw6761G1AddGas            uint64 = 600   // Price for BW6-761 elliptic curve G1 point addition
+	Bw6761G1MulGas            uint64 = 12000 // Price for BW6-761 elliptic curve G1 point scalar multiplication
+	Bw6761G2AddGas            uint64 = 4500  // Price for BW6-761 elliptic curve G2 point addition
+	Bw6761G2MulGas            uint64 = 55000 // Price for BW6-761 elliptic curve G2 point scalar multiplication
+	Bw6761PlonkProofVerifyGas uint64 = 21000 // Price for BW6-761 elliptic curve PlonK proof verification
```
Comment (iamrecursion) [inline params/protocol_params.go:150]: Given that gas does matter _somewhat_ just in terms of feasibility of computation, I would suggest working out roughly how many of the other precompiles _would_ be used in this and then summing their gas costs. Give or take a bit. I would expect it to certainly be the most expensive of all of these, and by quite a bit.

**Review by iamrecursion — APPROVED** [review]
One thing I do want to ask is whether it is worth us even holding onto the other precompiles? The less code that has to have potential collisions resolved when updating from upstream the better, and if we do not see any reason we're going to be using them in the future, we can remove them.

File: core/vm/contracts_ec.go (hunk 2 above)
Comment (wzmuda) [reply core/vm/contracts_ec.go:92]: The thing is we don't know too much about a byte-level contents. That code fully relies on Gnark putting stuff into a buffer and getting stuff out of a buffer and every step of that process depends on Gnark. The order is all we have control over. Also, whatever Gnark puts into the buffer, is already known to be unstable: I learned the hard way that a witness serialized by Gnark v0.9.2 is not recoverable by v0.9.1. If Gnark v.0.9.3 changes the alignment, order of fields or whatever, all we can do is to update Gnark used by Quorum.

Comment (iamrecursion) [reply core/vm/contracts_ec.go:92]: In that case I would be extremely clear about this, and how the Gnark version used by the proving service _must be the same_ as the one used by Quorum.

File: core/vm/contracts_ec.go [hunk 5]
```diff
@@ -73,6 +83,61 @@ func RunECPrecompiledContract(evm *EVM, precompile ECPrecompiledContract, input
...
 
+// bw6761PlonkProofVerifyPrecompile implements the verification of a given PlonK proof. It is assumed that the proof
+// was generated using BW6-761 curve.
+//
+// The input is assumed to encode, in the following order:
+// - a PlonK proof of a circuit,
+// - a verifying key for the proof,
+// - a public witness suitable for verifying the proof.
+//
+// All the above components must be generated for the same circuit definition, over the BW6-761 finite field.
```
Comment (wzmuda) [inline core/vm/contracts_ec.go:94]:
```suggestion
// All the above components must be generated for the same circuit definition, over the BW6-761 finite field.
//
// Gnark version used by the proving service MUST BE the same as the one used by Quorum. The binary representation
// of the serialized objects is Gnark's internal implementation detail. No assumptions can be made about the byte-level
// structure of the input buffer. The only way of making sure the buffer is structured correctly is to maintain the
// order of the objects and keep Gnark versions synchronized among the interfacing projects.
```

@iamrecursion that's regarding your previous comment

File: core/vm/contracts_ec.go (hunk 2 above)
Comment (wzmuda) [reply core/vm/contracts_ec.go:92]: I added a paragraph about that

File: params/protocol_params.go (hunk 4 above)
Comment (wzmuda) [reply params/protocol_params.go:150]: That should get revealed when I finally run the verification on my local chain. I'll then use the discovered value here.

Comment (wzmuda) [reply params/protocol_params.go:150]: Since there's an ACK for this PR, I'll merge it and open an issue for this gas value

Comment (wzmuda) [reply params/protocol_params.go:150]: I just realized I can't open an issue in this repo, there's no such option. I can only see the issues tab when I go into the upstream repo.

**Pull request merged**